source/2015-09-08-meteor-and-three-js.html.markdown

```diff
@@ -0,0 +1,230 @@
+---
+title: Meteor and THREE.js
+date: 2015-09-08 21:30
+tags:
+---
+
+We at PANTER already combined Meteor with a 2d-canvas-libraries ([Konva / Kinetic.js](https://github.com/konvajs/konva)), so 3d-rendering is the next logical step.
+
+## Introduction
+
+Meteor is a "fullstack" webframework, covering both client and server. If you haven't heard about it, check it out at [meteor.com](https://www.meteor.com/).
+
+Meteor update its views as soon as the underlying data changes in a reactive way. The default rendering engine *Blaze* and the template-language "spacebars" provide an easy way to work with html (btw. official *react*-support coming soon), but what if we want to use other renderers such as THREE.js and update a 3D-View in a reactive way?
+
+## Preparations
+
+I use the meteor-smartpackage 'limemakers:three' in this example, but you can add the THREE.js-files also manually.
+
+## Set the scene
+
+Let's start by defining a first template and attach data (a cursor) to it:
+
+~~~
+
+<template name="app">
+  <h1>my first 3d-App</h1>
+  {{>THREE_scene objects=players width="100%" height="500px"}}
+</template>
+
+Template.app.helpers
+  players: -> Collections.Players.find()
+
+~~~
+
+The *THREE_scene*-template defines a container which is later used by THREE.js to render the scene:
+
+~~~
+
+<template name="THREE_scene">
+  <div class="container" style="width: {{width}};height:{{height}}">
+  </div>
+</template>
+
+Template.THREE_scene.onRendered ->
+  # init renderer and scene
+  renderer = new THREE.WebGLRenderer antialias: true
+  renderer.setPixelRatio window.devicePixelRatio
+  renderer.setSize @$(".container").width(), @$(".container").height()
+  @$(".container").append renderer.domElement
+  scene = new THREE.Scene()
+
+  (...)
+
+~~~
+
+## Setup the camera
+
+Let's add a camera and make it rotateable with the mouse and let there be light:
+
+~~~
+
+Template.THREE_scene.onRendered ->
+
+  (...)
+
+  # init camera
+  camera = new THREE.PerspectiveCamera 75,
+    (@$(".container").width() / @$(".container").height()), 0.1, 1000
+  camera.position.z = 5
+  controls = new THREE.OrbitControls camera, renderer.domElement
+
+  # let there be light
+  scene.add new THREE.AmbientLight 0x505050
+  light = new THREE.SpotLight 0xffffff, 1.5
+  light.position.set 0, 500, 2000
+  light.castShadow = true
+  scene.add light
+
+  (...)
+
+~~~
+
+## start the rendering loop
+
+While the template is rendered, we can use *requestAnimationFrame* to re-render the scene. If we change any objects added to the scene, THREE.js will update all these objects.
+
+Notabene: we are still in the *Template.THREE_scene.onRendered*-callback. We can check if this template has been destroyed, by checking the *isDestroyed*-property on the underlying Blaze-view (ES2015-users may use the *fat-arrow*-function here, js-users keep a reference to the template with self=this).
```

Comment:
I would use `backticks` for all method names and other code. Also "Note that we..." instead of "Notabene", or "NB:" if you insist ;-)